### DISCUSSION TITLE
How to pass a list in a query string and handle it the endpoint function?

### DISCUSSION THREAD

**Discussion by igormorgado** [fastapi/fastapi#4185] [Questions] [answered]
Lists in query string are a fuzzy topic in HTTP specification. 

It can be seen in the links below:

https://stackoverflow.com/questions/24059773/correct-way-to-pass-multiple-values-for-same-parameter-name-in-get-request

https://stackoverflow.com/questions/724526/how-to-pass-multiple-parameters-in-a-querystring

https://en.wikipedia.org/wiki/Query_string

My question is: What is the correct (recommended) way to handle lists in GET query string? 

As an example these are the ways to make a query string list request

Method 1: `schema://host:port/path/?query=A&query=B&query=C` 

Method 2: `schema://host:port/path/?query=A,B,C` 

Method 3: `schema://host:port/path/?query[]=A&query[]=B&query[]=C` 

Are these cases handled by fastAPI? How can I handle then inside a endpoint?

```
@app.get("/path/")
def get_query(query: List[str]) -> Any:
    """This function handle a list of str in query"""
    # here goes the code
```

Naturally the *method2* can be handled considering query as a `string` and `str.split(',')` them.

My question are not related to this but if this is already done by fastAPI or if there are plans/roadmap to handle this.

In fact I would like to be able to handle the 3 cases at same time. The methods 1 and 3 will probably need internal fastAPI changes if these are not already handled.

Right now, in fastAPI 0.70, if I pass the type `List`  to a query parameter, fastAPI assumes that it will need a body, and that causes an error since GET methods do not handle requests with `body` .

**Comment by YuriiMotov** [comment, marked as answer]
For now FastAPI only supports 1st option (`?query=A&query=B&query=C`).

To make FastAPI treat this parameter as `Query` parameter you need to modify the signature of the endpoint function:
```
def get_query(query: Annotated[list[str], Query()]) -> Any:
```

Full code:
```python
from typing import Annotated, Any
from fastapi import FastAPI, Query


app = FastAPI()

@app.get("/path/")
def get_query(query: Annotated[list[str], Query()]) -> Any:
    return {"query": query}
```
